File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.py
class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        def union(parents, x, y):
            px, py = find(parents, x), find(parents, y)
            if(px != py):
                parents[py] = px
            
            return px == py
        
        def find(parents, x):
            if(parents[x] != x):
                parents[x] = find(parents, parents[x])
            
            return parents[x]
        
        def reachable(graph):
            prev = find(graph, 1)
            for i in range(2, n + 1):
                if(prev != find(graph, i)):
                    return False
            
            return True
        
        alice = {i:i for i in range(1, n + 1)}
        bob = {i:i for i in range(1, n + 1)}
        edges.sort(reverse=True)
        deleted = 0
        
        for edge_type, a, b in edges:
            if(edge_type == 3):
                deleted += union(alice, a, b) == \
                union(bob, a, b) == True
            elif(edge_type == 2):
                deleted += union(bob, a, b)
            else:
                deleted += union(alice, a, b)
        
        if(not reachable(alice) or not reachable(bob)):
            print(alice, bob)
            return -1
        
        return deleted
        
            

File: 1579-remove-max-number-of-edges-to-keep-graph-fully-traversable/1579-remove-max-number-of-edges-to-keep-graph-fully-traversable.go
type DisjointSet struct {
    parents map[int]int
}

func NewDisjointSet(n int) *DisjointSet {
	ds := &DisjointSet{
		parents: make(map[int]int),
	}
	for i := range n {
		ds.parents[i + 1] = i + 1
	}
	return ds
}

func (this *DisjointSet) Find(p int) int {
    var px = this.parents[p]
    if px != p {
        this.parents[p] = this.Find(px)
    }
    return this.parents[p]
}

func (this *DisjointSet) Union(a, b int) bool {
    var pa, pb = this.Find(a), this.Find(b)
    if pa != pb {
        this.parents[pb] = pa
    }
    return pa == pb
}

func maxNumEdgesToRemove(n int, edges [][]int) int {
    var aliceGraph = NewDisjointSet(n)
    var bobGraph = NewDisjointSet(n)
    var deleted = 0
    
    slices.SortFunc(edges, func(a, b []int) int{
        return cmp.Compare(b[0], a[0])
    })
    // Type 3 handler
    for _, edge := range edges {
        if edge[0] == 3 {
            var a, b = aliceGraph.Union(edge[1], edge[2]), bobGraph.Union(edge[1], edge[2])
            if a == true && b == true {
                deleted++
            }
        } else if (edge[0] == 1 && aliceGraph.Union(edge[1], edge[2])) {
            deleted++
        } else if (edge[0] == 2 && bobGraph.Union(edge[1], edge[2])) {
            deleted++
        }
    }

    var parentA, parentB = aliceGraph.Find(1), bobGraph.Find(1)
    for i := 2; i <= n; i++ {
        if parentA != aliceGraph.Find(i) || parentB != bobGraph.Find(i) {
            return -1
        }
    }
    
    return deleted
}